File: selling stocks/best time to buy and sell stock with fee.cpp
//************C++ Solution***********************

//memoization
//time: O(n) and space: O(n)
class Solution {
    int dp[50002][2];
public:
    int solve(vector<int> &prices, int fee, int pos, bool bought){
        if (pos == prices.size()){
            return 0;
        }
        if (dp[pos][bought] != -1){
            return dp[pos][bought];
        }
        if (bought){
            int buy = -prices[pos] + solve(prices, fee, pos + 1, false);
            int skip = solve(prices, fee, pos + 1, bought);
            return dp[pos][bought] =max(buy, skip);
        }
        else {
            int sell = (prices[pos] - fee) + solve(prices, fee, pos + 1, true);
            int skip = solve(prices, fee, pos + 1, bought);
            return dp[pos][bought] = max(sell, skip);
        }
    }
    int maxProfit(vector<int>& prices, int fee) {
        memset(dp, -1, sizeof(dp));
        return solve(prices, fee, 0, true);
    }
};

//************Python Solution***********************
class Solution:
    def solve(self, prices, fee, pos,bought,  dp):
        if pos == len(prices):
            return 0
        if dp[pos][bought] != -1:
            return dp[pos][bought]
        skip = self.solve(prices, fee, pos + 1, bought, dp)
        if bought:
            buy = - prices[pos] + self.solve(prices, fee, pos + 1, 0, dp);
            dp[pos][bought] = max(buy, skip);
            return dp[pos][bought]
        else:
            sell = (prices[pos] - fee) + self.solve(prices, fee, pos + 1, 1, dp)
            dp[pos][bought] = max(skip, sell);
            return dp[pos][bought]
    def maxProfit(self, prices: List[int], fee: int) -> int:
        dp = [[-1 for i in range(2)]for j in range(len(prices) + 1)]
        return self.solve(prices, fee, 0, 1, dp);
        

File: selling stocks/python solution/best time to buy and sell stock with fee.py
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        days = len(prices)
        profit = [[0 for i in range(2)] for j in range(days + 1)]
        for currDay in range(days - 1, -1, -1):
            for bought in range(2):
                skip = profit[currDay + 1][bought]
                if bought:
                    buy = -prices[currDay] + profit[currDay + 1][False]
                    profit[currDay][bought] = max(buy, skip)
                else:
                    sell = prices[currDay] - fee + profit[currDay + 1][True]
                    profit[currDay][bought] = max(sell, skip)
        return profit[0][True]
               
# space optimized solution
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        days = len(prices)
        ahead = [0, 0]
        current = [0, 0]
        for currDay in range(days - 1, -1, -1):
            for bought in range(2):
                skip = ahead[bought]
                if bought:
                    buy = -prices[currDay] + ahead[False]
                    current[bought] = max(buy, skip)
                else:
                    sell = prices[currDay] - fee + ahead[True]
                    current[bought] = max(sell, skip)
            ahead = current
        return current[True]
 # More clean code
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        days = len(prices)
        currentBuy = currentNotBuy = aheadBuy = aheadNotBuy = 0
        for currDay in range(days - 1, -1, -1):
            currentBuy = max(-prices[currDay] + aheadNotBuy, aheadBuy)
            currentNotBuy = max(prices[currDay] - fee + aheadBuy, aheadNotBuy)
            aheadBuy =  currentBuy
            aheadNotBuy = currentNotBuy
        return currentBuy
                    
